website/static/js/components/autocomplete.js

```diff
@@ -0,0 +1,189 @@
+ */
+var baseSearchViewModel = function (params) {
+    // Parse params
+    this.dataUrl = params.data;
+    this.inputElement = params.inputElement;
+
+    // Observables
+    this.items = ko.observableArray();
+    this.itemSelected = ko.observable(null);
+    this.hasItemSelected = ko.pureComputed(function() {
+        return this.itemSelected() !== null;
+    }.bind(this));
+    this.submitText = ko.observable(params.submitText || 'Submit');
+    this.placeholder = params.placeholder || '';
+
+    // Get the data and initialize the component
+    this.fetchData().done(
+        this.processData.bind(this),
+        this.initTypeahead.bind(this)
+    );
+};
+$.extend(baseSearchViewModel.prototype, {
+    /************************************************
+     * Abstract methods - subclasses must implement *
+     ************************************************/
+
+    /** Populate `items` from the `data` returned in `fetchData` */
+    processData: function(data) {
+        throw new MethodNotDefined('processData');
+    },
+    /**
+     * Given a list of typeahead items, return a function that accepts a query
+     *  and a callback, and calls the callback with a list of matched items.
+     */
+    substringMatcher: function(items) {
+        throw new MethodNotDefined('substringMatcher');
+    },
+    /** Given a typeahead item, return HTML to be displayed in the dropdown. */
+    suggestionTemplate: function() {
+        throw new MethodNotDefined('suggestionTemplate');
+    },
+    /** Called when the form is submitted */
+    onSubmit: function() {
+        throw new MethodNotDefined('onSubmit');
+    },
+    /** Get the data at the provided URL, return a promise */
+    fetchData: function () {
+        return $.get(this.dataUrl).fail(
+            function() {
+                console.log('failed to retrieve data');
+            });
+    },
+    /** Initialize the Typeahead widget for the component */
+    initTypeahead: function () {
+        this.inputElement.typeahead({
+            hint: false,
+            highlight: true,
+            minLength: 0
+        }, {
+            // name: 'projectSearch' + nodeType + namespace,
+            displayKey: function(data) {
+                return data.value.node.title;
+            },
+            templates: {
+                suggestion: this.suggestionTemplate
+            },
+            source: this.substringMatcher(this.items())
+        });
+
+        this.inputElement.bind('typeahead:selected', function(event, item) {
+            // Call the parent viewModel's onSelected
+            this.onSelected(item.value);
+        }.bind(this));
+    },
+    /** Clear the selection */
+    onClear: function () {
+        this.inputElement.typeahead('val', '');
+        this.itemSelected(null);
+    },
+    /** Set the item as selected */
+    onSelected: function (item) {
+        this.itemSelected(item);
+    }
+});
+
+
+var draftRegistrationsSearchViewModel = function (params) {
```

Comment:
Use `UpperCamelCase` for class names.

https://github.com/felixge/node-style-guide#use-uppercamelcase-for-class-names